desktop/src/main/backup.ts

```diff
@@ -0,0 +1,90 @@
+import client, { clientId } from '@hcengineering/client'
+import { getWorkspaceId, MeasureMetricsContext, type BackupClient, type Client } from '@hcengineering/core'
+import { addLocation, getResource, setMetadata } from '@hcengineering/platform'
+import WebSocket from 'ws'
+
+import {
+  backup,
+  createFileBackupStorage
+} from '@hcengineering/server-backup'
+import { dialog, type BrowserWindow } from 'electron'
+
+let runningBackup: {
+  notify: (command: string, ...args: any[]) => void
+} | undefined
+
+export async function createClient (
+  transactorUrl: string,
+  token: string
+): Promise<Client & BackupClient> {
+  // We need to override default factory with 'ws' one.
+  // eslint-disable-next-line
+  setMetadata(client.metadata.UseBinaryProtocol, true)
+  setMetadata(client.metadata.UseProtocolCompression, true)
+  setMetadata(client.metadata.ConnectionTimeout, 0)
+  setMetadata(client.metadata.ClientSocketFactory, (url) => {
+    return new WebSocket(url) as any
+  })
+  addLocation(clientId, () => import('@hcengineering/client-resources'))
+
+  const clientFactory = await getResource(client.function.GetClient)
+  return (await clientFactory(token, transactorUrl)) as unknown as Client & BackupClient
+}
+
+async function doBackup (dirName: string, token: string, endpoint: string, workspace: string, notify: (command: string, ...args: any[]) => void): Promise<void> {
+  notify('backup', 0)
+  const ctx = new MeasureMetricsContext('backup', {})
+
+  const storage = await createFileBackupStorage(dirName)
+  const wsid = getWorkspaceId(workspace)
+  const client = await createClient(endpoint, token)
+  try {
+    await backup(ctx, endpoint, wsid, storage, {
+      force: true,
+      recheck: false,
+      skipDomains: [],
+      timeout: 0,
+      connectTimeout: 30 * 1000,
+      blobDownloadLimit: 15,
+      token,
+      skipBlobContentTypes: [],
+      isCanceled: (): boolean => {
+        return runningBackup == null
+      },
+      progress: (value: number): void => {
+        notify('backup', value)
+      },
+      connection: client
+    })
+  } finally {
+    await client.close()
+  }
+
+  notify('backup', 100)
+}
+
+export function startBackup (window: BrowserWindow, token: string, endpoint: string, workspace: string, cmd: (command: string, ...args: any[]) => void): void {
+  void dialog
+    .showOpenDialog(window, {
+      properties: ['openDirectory'],
+      buttonLabel: 'Backup...',
+      message: 'Select a folder for incremental backup. Huly will will  download all new changes from workspace.'
```

Comment:
Do we need to check that it was the same workspace in case of incremental backup?

Reply:
Nope, it will do a file diff hash, so it will assume one was fully migrated to another. And since it is incremental it will be possible to restore to any of intermediate states.